assets/js/hook.js

```diff
@@ -0,0 +1,270 @@
+////////////////////////////////////////////////////////////////////////////////////////
+// Hook
+////////////////////////////////////////////////////////////////////////////////////////
+class Hook {
+    constructor(id, node, caption = '', type = 'default', io = 'out',
+                links_options = {}, expects = null, provides = null) {
+        /**
+         * @param {Object} [options]
+         * @param {string} [options.id] - hook id.
+         * @param {fabric.Node} [options.node] - node that this hook belongs to
+         * @param {string} [options.caption] - caption to print in front of the hook
+         * @param {string} [options.io] - 'in' or 'out', defines if this hook is input or output of its node
+
+         */
+        this.node = node;
+        this.caption = caption;
+        this.type = type;
+        this.io = io;
+        this.id = id || this.generateID(io); //TODO check if unique
+
+        this.edges = [];
+        this.links = {};
+        this.links_options = links_options;
+        this.provides = provides;
+        this.expects = expects;
+        this.graph = node.graph
+    }
+
+    generateID(io) {
+        return io + '_' + (this.node.hooks[io].length + 1);
+    }
+
+    neighborhood(point, radius = 10) {
+        let center = this.getCenter();
+
+        return (Math.abs(center.x - point.x) < radius
+            && Math.abs(center.y - point.y) < radius);
+    }
+
+    allowConnection(hook) {
```

Comment:
Because, the way we designed the stuff, the edge/connection is carried by both source and target hooks. But we don't want to check twice if the connection is legal. Hence, it's enough to check only one direction, and because the endoint is attached to the output node, it appeared to me quite fair to check only when `this.io === 'in'`. 
Do you think of a use-case where it would fail?